### ISSUE TITLE
Strange Python 3 behavior after setting attributes

### ISSUE THREAD

**Issue by ghackebeil** [pympler/pympler#53]
When I define a simple class such as

```python
class block(object):
    pass
```
and then print the size of an instance of that class (using `pympler.asizeof.asizeof`), I get roughly equivalent behavior on Python 2.7.13 (size=344 B) and Python 3.6.1 (size=168 B).

However, if I print the size of an instance of that class after adding a large number of attributes **to an entirely separate instance of that class**, I see very strange behavior with Python 3. In particular, the size of the new, and supposedly empty, instance of that class is huge (almost 100 KB). Is this a bug or am I using `asizeof` wrong?

Here is a complete example:

```python
import pympler.asizeof

def _fmt(num, suffix='B'):
    """format memory output"""
    for unit in ['','K','M','G','T','P','E','Z']:
        if abs(num) < 1000.0:
            return "%3.1f %s%s" % (num, unit, suffix)
        num /= 1000.0
    return "%.1f %s%s" % (num, 'Yi', suffix)

class block(object):
    pass

def f():
    b = block()
    for i in range(10000):
        setattr(b, "c"+str(i), i)

print(_fmt(pympler.asizeof.asizeof(block())))
f()
print(_fmt(pympler.asizeof.asizeof(block())))
```

For Python 2.7.13, this outputs:
```bash
344.0 B
344.0 B
```
For Python 3.6.1, this outputs:
```bash
168.0 B
87.5 KB
```

**Comment by mrJean1** [comment]
With Python 3.6.1 on macOS 10.12.5, the first block() instance is 168 bytes, but all subsequent sizes are 296 bytes.  Not as bad as 87.5 K, but still unexpected.

**Comment by ghackebeil** [comment]
I'm on the same OS. Python3 was installed via Homebrew. Strange that we get different sizes.

**Comment by ghackebeil** [comment]
Interestingly, if `setattr(b, "c"+str(i), i)` is replaced with `b.__dict__["c"+str(i)] = i`, there is no increase in memory.

**Comment by Pankrat** [comment]
I can reproduce this on Linux (Ubuntu 16.04) with Python 3.6 and 3.5.

Asizeof reports a much larger flatsize of the `__dict__` attribute even though there's no data stored in the dict: 

```
>>> print(asizeof.asized(with_attributes, detail=1, code=False).format())
block object at 0x7f4964ebddd8> size=99040 flat=56
    __dict__ size=98984 flat=10992
    __class__ size=0 flat=0
>>> print(asizeof.asized(without_attributes, detail=1, code=False).format())
block object at 0x7f4964ebde48> size=11048 flat=56
    __dict__ size=10992 flat=10992
    __class__ size=0 flat=0
```

I'll try to dig a bit deeper.

**Comment by mrJean1** [comment]
One other data point from the original bug report.

print(_fmt(asizeof.asizeof(block(), stats=0)))
f()
print(_fmt(asizeof.asizeof(block(), stats=0)))

… reproduces the problem:

% python2.6.9 test.py
344.0 B
344.0 B
% python2.7.14 test.py
344.0 B
344.0 B
% python3.5.3 test.py
152.0 B
131.2 KB
% python3.6.4 test.py
168.0 B
87.5 KB


However, the following variant ...

b = block()
print(_fmt(asizeof.asizeof(b)))
f()
a = block()
print(_fmt(asizeof.asizeof(a)))

… does not reproduce the problem:

% python2.6.9 test.py
344.0 B
344.0 B
% python2.7.14 test.py
344.0 B
344.0 B
% python3.5.3 test.py
152.0 B
344.0 B
% python3.6.4 test.py
168.0 B
296.0 B

Still digging, /Jean

**Comment by mrJean1** [comment]
Even more unusual behavior after adding a few print statements.

def f():
    b = block()
    print('f:', b)
    for i in range(10000):
        setattr(b, "c"+str(i), i)

print('b:', block())
print(_fmt(asizeof.asizeof(block(), stats=0)))
f()
print('a:', block())
print(_fmt(asizeof.asizeof(block(), stats=0)))

All block() instances all have the same id, memory address.  That is not entirely unexpected, since the memory freed from a previous instance may be reused for a future instance.

% python2.6.9 test.py
('b:', <main__.block object at 0x10d77c750>)
344.0 B
('f:', <__main__.block object at 0x10d77c750>)
('a:', <__main__.block object at 0x10d77c750>)
344.0 B

% python2.7.14 test.py
('b:', <__main__.block object at 0x10060cd90>)
344.0 B
('f:', <__main__.block object at 0x10060cd90>)
('a:', <__main__.block object at 0x10060cd90>)
344.0 B

% python3.5.3 test.py
b: <__main__.block object at 0x10181c1d0>
152.0 B
f: <__main__.block object at 0x10181c208>
a: <__main__.block object at 0x10181c208>
131.2 KB

% python3.6.4 test.py
b: <__main__.block object at 0x1018c8048>
168.0 B
f: <__main__.block object at 0x1018c8080>
a: <__main__.block object at 0x1018c8080>
87.5 KB


However, when ignoring the block() instance after the f() call as follows ...

def f():
    b = block()
    print('f:', b)
    for i in range(10000):
        setattr(b, "c"+str(i), i)

b = block()
print('b:', b)
print(_fmt(asizeof.asizeof(b, stats=0)))
f()
_ = block()  # ignore this instance
a = block()
print('a:', a)
print(_fmt(asizeof.asizeof(a, stats=0)))

… then all instance are different and are sized correctly.

% python2.6.9 test.py
('b:', <__main__.block object at 0x103a99b50>)
344.0 B
('f:', <__main__.block object at 0x103ab5110>)
('a:', <__main__.block object at 0x103ab5390>)
344.0 B

% python2.7.14 test.py
('b:', <__main__.block object at 0x100646d90>)
344.0 B
('f:', <__main__.block object at 0x1033132d0>)
('a:', <__main__.block object at 0x103313310>)
344.0 B

% python3.5.3 test.py
b: <__main__.block object at 0x10181d160>
152.0 B
f: <__main__.block object at 0x10181d240>
a: <__main__.block object at 0x101c4a320>
344.0 B

% python3.6.4 test.py
b: <__main__.block object at 0x1018c8048>
168.0 B
f: <__main__.block object at 0x1018c80f0>
a: <__main__.block object at 0x101bf34a8>
296.0 B

In summary, this looks like an issue with the Python 3+ memory manager.

/Jean

**Comment by Pankrat** [comment]
Thanks Jean, I get the same results on my box. 

My other tests also indicate that this is either a feature or a bug of the Python 3 memory manager. This behavior can be reproduced without pympler by just using `sys.getsizeof(f.__dict__)`:

```
>>> sys.getsizeof(f.__dict__)
10992
>>> f.__dict__
{}
```

Interestingly this size keeps showing on copies of the dict as well:

```
>>> sys.getsizeof(f.__dict__.copy())
10992
```

I believe the question for us is whether Python truly allocates this much memory for the instances. If it does, the profiling is correct. If it doesn't, this would be a bug in the implementation of CPython's `sys.getsizeof`.

**Comment by mrJean1** [comment]
Thank you.  It is a Python 3 issue, I believe, the memory manager re-uses freed memory without clearing or initializing it entirely.

And we should report it as a Python 3.5 and 3.6 bug using sys.getsizeof not asizeof.  

/Jean

**Comment by Pankrat** [comment]
Here's my test case:

```
class Empty(object):
    pass

a = Empty()
print("{} {}".format(id(a.__dict__), sys.getsizeof(a.__dict__)))
for i in range(1000):
    setattr(a, "c"+str(i), i)
b = Empty()
print("{} {}".format(id(b.__dict__), sys.getsizeof(b.__dict__)))
```

This issue affects all Python 3.x versions back to Python 3.3:

```
$ python3.6 test53.py 
139862212160392 112
139862211472408 10992
$ python3.5 test53.py 
140009284245448 96
140009284245832 16448
$ python3.4 test53.py 
140457678079048 96
140457677927880 16448
$ python3.3 test53.py 
139988731797304 104
139988707431992 16456
$ python3.2 test53.py 
11518832 280
12700512 280
$ python2.7 test53.py 
139623596262576 280
139623596263136 280
```

As @ghackebeil mentions, this doesn't happen when modifying `__dict__` directly, only when using `setattr`.

**Comment by mrJean1** [comment]
This looks good, actually this shows the problem is even worse, more specific. Because (1) instance b is affected by instance a and (2) it only happens with setattr.

Only one comment, add

import sys

/Jean

**Comment by Pankrat** [comment]
Looks like this is a profiling error, not an allocation pattern. Here's a test that measures the process size after instantiating a large quantity of objects:

```python
import sys
from pympler import asizeof
from pympler import process
from pympler.process import ProcessMemoryInfo

class Empty(object):
    pass

print("Test sizing objects without attributes...")
z = Empty()
a = Empty()
print("Empty dict size:          ID={} {}".format(id({}), sys.getsizeof({})))
print("Dict size before setattr: ID={} {}".format(id(a.__dict__), sys.getsizeof(a.__dict__)))
for i in range(10000):
    setattr(a, "c"+str(i), i)

pmi = ProcessMemoryInfo()
before = pmi.vsz

b = Empty()
c = Empty()
print("Dict size after setattr:  ID={} {}".format(id(a.__dict__), sys.getsizeof(a.__dict__)))
print("Dict size other instance: ID={} {}".format(id(b.__dict__), sys.getsizeof(b.__dict__)))
print("Dict size third instance: ID={} {}".format(id(c.__dict__), sys.getsizeof(c.__dict__)))
print("Dict size predecessor:    ID={} {}".format(id(z.__dict__), sys.getsizeof(z.__dict__)))
print("\nInstantiating 100000 instances ...")
many = [Empty() for _ in range(100000)]

pmi.update()
print("VSZ delta:                {} KB".format((pmi.vsz - before) // 1000))
print("Recursive size:           {} KB".format(asizeof.asizeof(many) // 1000))
print("VSZ per instance:         {} B".format((pmi.vsz - before) // 100000))
```

And the output on my box:

```
Test sizing objects without attributes...
Empty dict size:          ID=140057526516616 240
Dict size before setattr: ID=140057526516616 112
Dict size after setattr:  ID=140057526516616 87448
Dict size other instance: ID=140057495989272 87448
Dict size third instance: ID=140057495988984 87448
Dict size predecessor:    ID=140057495988768 87448

Instantiating 100000 instances ...
VSZ delta:                6844 KB
Recursive size:           8751224 KB
VSZ per instance:         68 B
```

So roughly 68 bytes were allocated for each additional instance. If the flatsize of 87448 bytes was correct, this number should be much bigger.

This only happens when calling `setattr` on objects without attributes. If you set an attribute on any instance before or after calling setattr, the sizes appear to be correct. It looks like Python chooses a more compact format for attribute storage if there are no attributes set on instances of a class. 

I suspect a copy-on-write optimization which doesn't work correctly (for `sys.getsizeof`) when using `setattr`.

**Comment by mrJean1** [comment]
This all seem correct.  The flat sizes of the b and c instances are clearly wron, obviously.  But the average size of 68 bytes includes the Empty instance plus a reference in the many list of 8+ bytes.  On my Mac, the asizeof (== sys.getsizeof) an Empty instance is 56 bytes and an empty dict() is 240 bytes (using 64-bit Python 3.6.4).  With 64-bit Python 2.6.9 those sizes are 344 respectively 280 bytes. 

/Jean
